misk-redis/src/main/kotlin/misk/redis/RealRedis.kt

```diff
@@ -30,20 +33,64 @@ class RealRedis(
   }
 
   override fun del(vararg keys: String): Int {
-    val keysAsBytes = keys.map { it.toByteArray(charset) }.toTypedArray()
-    return jedis { del(*keysAsBytes) }.toInt()
+    return when(unifiedJedis) {
+      is JedisPooled -> {
+        val keysAsBytes = keys.map { it.toByteArray(charset) }.toTypedArray()
+        jedis { unifiedJedis.del(*keysAsBytes) }.toInt()
+      }
+
+      is JedisCluster -> {
+        // JedisCluster does not support multi-key del, so we need to group by slot and perform del for each slot
+        keys.groupBy { JedisClusterCRC16.getSlot(it) }
+          .map { (_, slotKeys) -> jedis { unifiedJedis.del(*slotKeys.toTypedArray()) } }
+          .sumOf { it.toInt() }
+      }
+
+      else -> throw RuntimeException("Unsupported UnifiedJedis implementation ${unifiedJedis.javaClass}")
+    }
   }
 
   override fun mget(vararg keys: String): List<ByteString?> {
-    val keysAsBytes = keys.map { it.toByteArray(charset) }.toTypedArray()
-    return jedis { mget(*keysAsBytes) }
-      .map { it?.toByteString() }
+    return when(unifiedJedis) {
+      is JedisPooled -> {
+        val keysAsBytes = keys.map { it.toByteArray(charset) }.toTypedArray()
+        jedis { unifiedJedis.mget(*keysAsBytes) }.map { it?.toByteString() }
+      }
+      is JedisCluster -> {
+        // JedisCluster does not support multi-key mget, so we need to group by slot and perform mget for each slot
+        val keyToValueMap = mutableMapOf<String, ByteString?>()
+        keys.groupBy { JedisClusterCRC16.getSlot(it) }
+          .flatMap { (_, slotKeys) ->
+            val result = jedis { unifiedJedis.mget(*slotKeys.toTypedArray()) }
+            slotKeys.zip(result)
+        }.forEach { (key, value) ->
+          keyToValueMap[key] = value?.toByteArray(charset)?.toByteString()
+        }
+        keys.map{keyToValueMap[it]}
+      }
+
+      else -> throw RuntimeException("Unsupported UnifiedJedis implementation ${unifiedJedis.javaClass}")
+    }
+
   }
 
   override fun mset(vararg keyValues: ByteString) {
     require(keyValues.size % 2 == 0) { "Wrong number of arguments to mset" }
-    val byteArrays = keyValues.map { it.toByteArray() }.toTypedArray()
-    jedis { mset(*byteArrays) }
+    when(unifiedJedis) {
+      is JedisPooled -> {
+        val byteArrays = keyValues.map { it.toByteArray() }.toTypedArray()
+        return jedis { unifiedJedis.mset(*byteArrays) }
+      }
+      is JedisCluster -> {
+        // JedisCluster does not support multi-key mset, so we need to group by slot and perform mset for each slot
+        keyValues.toList().chunked(2).groupBy { JedisClusterCRC16.getSlot(it[0].toByteArray()) }
+          .forEach { (_, slotKeys) ->
+            jedis { unifiedJedis.mset(*slotKeys.flatten().map{it.toByteArray()}.toTypedArray())}
```

Comment:
Should we say something in the docs that this can trigger severals calls to redis?

Reply:
Updated